# Linear Regression Baseline

This notebook implements a baseline Linear Regression model for predicting hourly bike demand.

In [1]:
import pandas as pd
import numpy as np
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_absolute_error, mean_squared_error

## Load and Clean Data

We load the processed dataset and remove rows with missing values caused by lag features.

In [3]:
df = pd.read_csv("../data/processed/final_dataset.csv")
df = df.dropna()

df.head()

,station_id,hour,weekday,month,lag_1,lag_24,member_ratio,is_weekend,historical_avg_demand,trips
24,7000,5,1,1,1.0,10.0,0.500000,False,49.68652,4
25,7000,6,1,1,4.0,6.0,0.052632,False,49.68652,19
26,7000,7,1,1,19.0,16.0,0.217391,False,49.68652,23
27,7000,8,1,1,23.0,19.0,0.211268,False,49.68652,71
28,7000,9,1,1,71.0,42.0,0.285714,False,49.68652,28


## Train/Test Split

We use a time-based split:
- First 80% → training set  
- Last 20% → test set

In [4]:
split_idx = int(len(df) * 0.8)

train = df.iloc[:split_idx]
test = df.iloc[split_idx:]

## Prepare Features and Train Model

We use all features except the target (`trips`) to train a Linear Regression model.

In [6]:
X_train = train.drop(columns=["trips"])
y_train = train["trips"]

X_test = test.drop(columns=["trips"])
y_test = test["trips"]

model = LinearRegression()
model.fit(X_train, y_train)

print(model)

LinearRegression()


## Evaluation

We evaluate the model using MAE and RMSE.

In [7]:
y_pred = model.predict(X_test)

mae = mean_absolute_error(y_test, y_pred)
rmse = np.sqrt(mean_squared_error(y_test, y_pred))

print("MAE:", mae)
print("RMSE:", rmse)

MAE: 3.8258618783491776
RMSE: 6.875281945729522


## Results

- MAE ≈ 3.83
- RMSE ≈ 6.88

## Analysis

The Linear Regression model captures general demand patterns,  
but its performance is limited due to its linear assumptions.

More advanced models such as Random Forest and XGBoost are expected to perform better.